### PR TITLE
graph/encoding/{,di}graph6: fix bit handling for graph size encoding

### LINKED ISSUE #1616: graph/encoding/{,di}graph6: Encode produces invalid encoding that is missed by IsValid
Encoding the Sudoko board graph results in the following encoding (note the non-ASCII character at position 4):
```
~?@~~~~~~wF?{BFFFFbbwF~?~{B~~?wF?wF_[BF?wwwFFb_[^?wF~?wF~_[B~{?_C?OA?OC_C?_W_C?_wOA?O{C?_C^?_C?fwA?OA~_C?_F~_C?_F?OA?OF?c?_CB_W_C?_FFA?OA?w{C?_CBbwC?_C?~wA?OA?~{?_C?_^~_C?_F?wA?OA?wF?c?_CB_[BC?_C?wFFA?OA?wFF__C?_[BbwC?_C?wF~?OA?OF?~{?_C?_[B~{?_C?_C?_A?OA?OA?OC_C?_C?_CBC?_C?_C?_wOA?OA?OAF__C?_C?_C^?_C?_C?_C~?OA?OA?OA~_C?_C?_C?~{?_C?_C?_F?OA?OA?OA?wC_C?_C?_CB_W_C?_C?_C?wwOA?OA?OA?w{C?_C?_C?_[^?_C?_C?_C?~wA?OA?OA?OF~_C?_C?_C?_^~_C?_C?_C?wF?OA?OA?OA?wF?c?_C?_C?_[B_W_C?_C?_C?wFFA?OA?OA?OF?w{C?_C?_C?_[BbwC?_C?_C?_F?~wA?OA?OA?OF?~{?_C?_C?_CB_^~
```

The reference parser, [showg](https://users.cecs.anu.edu.au/~bdm/data/showg.c) errors when trying to read this.
```
$ showg -e sudoku.g6 
>E readgg: illegal character
```

This is due to carelessness in handling the bits. A PR is coming.

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Fixes #1616.

### CODE REVIEW

**Comment by kortschak** [comment]
The basis for this is that each byte of the encoding is a printable ASCII character (63-126), starting from a value 0-62. I had failed to mask against 0b111111 before writing to the buffer for values over 62 while doing the repeated 6-bit shifts (forgetting that bytes are 8 bits wide!).

File: graph/encoding/graph6/graph6_test.go [hunk 1]
```diff
@@ -217,3 +218,46 @@ func linksTo(nodes ...graph.Node) map[int]struct{} {
...
+	} {
+		g6 := Encode(implicitCycle(l))
+		if !IsValid(g6) {
+			t.Errorf("graph6-encoding unexpectedly invalid: %v", g6)
+		}
+		for i, b := range []byte(g6) {
+			if b < 63 || 126 < b {
+				t.Errorf("graph6-encoding contains invalid character at %d: %q", i, b)
+			}
+		}
```
Comment (mattConn) [inline graph/encoding/graph6/graph6_test.go:234]: `numberOf()`  performs this check in `IsValid()`. Should `IsValid` and `numberOf` return additional errors for better reporting? Additional checks like these would then no longer need to be performed.

**Review by mattConn — APPROVED** [review]
My graph6 remarks also apply to digraph6. I think more comments relating this implementation to the linked text could be helpful to the next reader. Looks good to me.

File: graph/encoding/graph6/graph6.go [hunk 2]
```diff
@@ -90,6 +90,10 @@ func Encode(g graph.Graph) Graph {
 	return Graph(buf.String())
 }
 
+func bit6(b int64) byte {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/encoding/graph6/graph6.go:93]: This is not really doing the splitting, but I'll add a comment.

File: graph/encoding/graph6/graph6.go [hunk 3]
```diff
@@ -67,9 +67,9 @@ func Encode(g graph.Graph) Graph {
...
 		buf.WriteByte(byte(n) + 63)
 	case n < 258048:
-		buf.Write([]byte{126, byte(n>>12) + 63, byte(n>>6) + 63, byte(n) + 63})
+		buf.Write([]byte{126, bit6(n>>12) + 63, bit6(n>>6) + 63, bit6(n) + 63})
 	case n < 68719476736:
-		buf.Write([]byte{126, 126, byte(n>>30) + 63, byte(n>>24) + 63, byte(n>>18) + 63, byte(n>>12) + 63, byte(n>>6) + 63, byte(n) + 63})
+		buf.Write([]byte{126, 126, bit6(n>>30) + 63, bit6(n>>24) + 63, bit6(n>>18) + 63, bit6(n>>12) + 63, bit6(n>>6) + 63, bit6(n) + 63})
 	default:
 		panic("graph6: too large")
 	}
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/encoding/graph6/graph6.go:75]: Done.

File: graph/encoding/graph6/graph6_test.go (hunk 1 above)
Comment (kortschak) [reply graph/encoding/graph6/graph6_test.go:234]: No, I don't think so. The API as it is is nicely light, having additional things happening in the tests to support that is OK.

**Review by mattConn — APPROVED** [review]
Looks good to me.

**Comment by kortschak** [comment]
Thank you for the review.

**Pull request merged**